### PR TITLE
Add 2D plotting script PostFit2DShapesFromWorkspace.cpp

### PR DESCRIPTION
**Pull request by mroguljic**
Resurrection of [this PR](https://github.com/cms-analysis/CombineHarvester/pull/223). I just checked that it compiles on el9.
@ajgilbert your comment on code duplication still stands, but it has been some time since the linked PR and I'm not sure if these functionalities have been added.

### CODE REVIEW

File: CombineTools/src/CombineHarvester.cc [hunk 1]
```diff
@@ -483,10 +483,21 @@ void CombineHarvester::LoadShapes(Process* entry,
...
           std::string var_name;
-          if (data_obj) var_name = data_obj->get()->first()->GetName();
+          // Custom code start - LC 5/14/19
+          std::string var_name_y;
+          if (data_obj) {
+            var_name = data_obj->get()->first()->GetName();
+            RooArgSet* temp_vars = (RooArgSet*)data_obj->get()->Clone();
+            RooAbsArg* temp_xvar = data_obj->get()->first();
+            temp_vars->remove(*temp_xvar,true,true);
+            var_name_y = temp_vars->first()->GetName();
```
Comment (ajgilbert) [inline CombineTools/src/CombineHarvester.cc:493]: Do we not need to check the data object actually has two observables first? Won't this crash if we only have one?

File: CombineTools/src/CombineHarvester.cc [hunk 2]
```diff
@@ -483,10 +483,21 @@ void CombineHarvester::LoadShapes(Process* entry,
...
+          }
+          // if (data_obj) var_name = data_obj->get()->first()->GetName();
           entry->set_observable(
               (RooRealVar*)entry->pdf()->findServer(var_name.c_str()));
-        }
+          entry->set_observable_y(
```
Comment (ajgilbert) [inline CombineTools/src/CombineHarvester.cc:498]: As above, surely we only want to call this if there is a 2nd observables?

File: CombineTools/interface/Process.h [hunk 3]
```diff
@@ -79,6 +84,7 @@ class Process : public Object {
   RooAbsData* data_;
   RooAbsReal* norm_;
   RooRealVar* cached_obs_;
+  RooRealVar* cached_obsy_;
```
Comment (ajgilbert) [inline CombineTools/interface/Process.h:87]: In Process.cc, need to handle this new data member correctly in all constructors, assignment and swap methods.

File: CombineTools/src/CombineHarvester_Evaluate.cc [hunk 4]
```diff
@@ -758,6 +1069,39 @@ void CombineHarvester::SetPdfBins(unsigned nbins) {
...
+  for (unsigned i = 0; i < procs_.size(); ++i) {
+    std::set<std::string> binning_varsx;
+    std::set<std::string> binning_varsy;
+    if (procs_[i]->pdf()) {
+      RooAbsData const* data_obj = FindMatchingData(procs_[i].get());
+      std::string var_name = "CMS_th1x";
+      std::string var_name_y = "CMS_th1y";
+      if (data_obj) {
+        var_name = data_obj->get()->first()->GetName();
+        RooArgSet* temp_vars = (RooArgSet*)data_obj->get()->Clone();
```
Comment (ajgilbert) [inline CombineTools/src/CombineHarvester_Evaluate.cc:1082]: Check memory ownership of cloned object

**Review by ajgilbert — CHANGES_REQUESTED** [review]
Hi,

The code introduces a lot of duplication, which is not ideal. That said, it can be merged pending a number of issues (see the code comments for details).

**Comment by mroguljic** [comment]
Thank you for the review @ajgilbert 
The PR was updated per your comments. I tested it on my usual 2D workflows, do you have some workflows that I could run to ensure it does not break the typical 1D case?

**Comment by ajgilbert** [comment]
Thanks, I think you could run on any set of 1D cards. For example there are some in the tutorials section: https://github.com/cms-analysis/HiggsAnalysis-CombinedLimit/tree/main/data/tutorials/htt/125

**Review by mroguljic — COMMENTED** [review]
Implemented the requested changes:
- add a flag to ensure that we run 2D part of the code only if we have more than one observables
- delete temporarily needed objects
- update Process.cc to properly handle new class members

**Review by ajgilbert — APPROVED** [review]

**Comment by lcorcodilos** [comment]
🎉 Thanks for getting this over the line @mroguljic and @ajgilbert! And glad to see it's still providing value 😄

**Pull request merged**